### ISSUE TITLE
[HOLD for payment 2023-06-13] [$1000]  In the assign task flow, Value for description not changing

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#19460]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Click on Assign Task in global create menu.
2. Add title and description.
3. Click on the next button.
4. On Confirm Task page, click on the Description
5. Change the value for description
6. Click on Next. You will go to Confirm Task menu
7. Click on back chevron button near heading
8. Description is not updated

## Expected Result: 
Description should be updated
## Actual Result: 
Description is not updated
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.17.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/044a7528-331d-44ec-a1ff-929e374ac805

https://github.com/Expensify/App/assets/43996225/805b89b5-ba0e-446d-98fb-0ad462bc8fea

**Expensify/Expensify Issue URL:**
**Issue reported by:** @BhuvaneshPatil
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1684554959081519

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0159a2a911d29c5676</li>
        <li>Upwork Job ID: 1661123057838981120</li>
        <li>Last Price Increase: 2023-05-30</li>
    </ul>
</details>

**Comment by greg-schroeder** [comment]
Potentially related https://github.com/Expensify/App/issues/19343

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
In the assign task flow, the value for description on in the initial page is not updated if it is changed on the details page and the user navigates back to the initial page.

### What is the root cause of that problem?
The taskDescription input field in the form on the initial page is not being updated with the stored value in Onyx.

### What changes do you think we should make in order to solve the problem?
Update the NewTaskDetailsPage to retrieve the value from Onyx when the page loads.

### What alternative solutions did you explore? (Optional)
None.

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
In the assign task flow, the value for description on in the initial page is not updated if it is changed on the details page and the user navigates back to the initial page.

### What is the root cause of that problem?
The taskDescription input field in the form on the initial page is not being updated with the stored value in Onyx.

### What changes do you think we should make in order to solve the problem?
Update the NewTaskDetailsPage to retrieve the value from Onyx when the page loads.

### What alternative solutions did you explore? (Optional)
None.

**Comment by dukenv0307** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
In the assign task flow, value for title and description not changing
### What is the root cause of that problem?
In `NewTaskDetailPage`, title and description input  field don't have `value` prop to update with task value from Onyx
https://github.com/Expensify/App/blob/7a42755a79737314edf702ce17bf0647dd8c41fb/src/pages/tasks/NewTaskDetailsPage.js#L78-L91
### What changes do you think we should make in order to solve the problem?
1. Subscribe task from Onyx with key `ONYXKEYS.TASK` and update `propTypes` and `defaultProps` corresponding
2. Add `value` prop for title field `value={props.task.title}` and description field `value={props.task.description}`

### What alternative solutions did you explore? (Optional)
NA

### Result

[Screencast from 25-05-2023 13:01:10.webm](https://github.com/Expensify/App/assets/129500732/fe9b0151-8822-4004-943d-fe11c7ea1b52)

**Comment by greg-schroeder** [comment]
@abdulrahuman5196 thoughts on the proposals above?

**Comment by tienifr** [comment]
### Proposal

### Please re-state the problem that we are trying to solve in this issue.

The title and description are not updated in AssignTask page

### What is the root cause of that problem?

In 

https://github.com/Expensify/App/blob/main/src/pages/tasks/NewTaskPage.js

We're using Form.js to render the form, it's data is only updated when user changes value manually. That why when we update title or description from other forms and go back, this form still shows stale data

### What changes do you think we should make in order to solve the problem?
- add new field `shouldSync` in task
- When we call `setTitleValue` or `setDescriptionValue` function, update shouldSync to true
- In NewTaskDetailPage subscribe `task` from onyx, if shouldSync is true -> update inputValues and reset shouldSync to false

```
    useEffect(()=>{
        if(props.task.shouldSync){
            formRef.current.setState({
                inputValues:{
                    taskTitle:props.task.title,
                    taskDescription: props.task.description
                }
            })
            TaskUtils.setShouldSync(false)
        }
    },[props.task.shouldSync])
```

### Result

https://github.com/Expensify/App/assets/113963320/5625f5db-bc0c-4f6a-979d-2f77c2e17b36

**Comment by BhuvaneshPatil** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue.
Title and Description value in `Assign Task` form not updating, after we change there values in next pages.

### What is the root cause of that problem?
There are two possible reasons for these - 
- The `NewTaskPage` components does not take any prop from onyx's task. That's why updating/re-rendering is not happening when those values are changed outside of that component.
https://github.com/Expensify/App/blob/0070968b27dd29728d5bb13a8b7eabb3db37a5ff/src/pages/tasks/NewTaskDetailsPage.js#L26-L28
- The `Form` component used in above mentioned component doesn't take any default values as of now, and if we check component tree it's not unmounted even if we go to another page i.e. `Confirm Task`. So the values that are present in both input fields are only particular to that component only.
https://github.com/Expensify/App/blob/0070968b27dd29728d5bb13a8b7eabb3db37a5ff/src/pages/tasks/NewTaskDetailsPage.js#L70-L77

### What changes do you think we should make in order to solve the problem?
There are few changes that we need to make - 
1. Add `task` to props, that will subscribe to the task from onyx.
2. The current implementation of `Form` component uses `draftValues` as props, they are used to initialize the `state.inputValues`, we can use them for our use case.
https://github.com/Expensify/App/blob/0070968b27dd29728d5bb13a8b7eabb3db37a5ff/src/components/Form.js#L94-L99
3. To make sure we are updating the values for the `Form` component, we can modify - `componentDidUpdate()` in `Form` component like following - 
```
    componentDidUpdate(prevProps) {
      if (!_.isEqual(prevProps.draftValues, this.props.draftValues)) {
          this.setState((pre) => ({...pre, inputValues: {...pre.inputValues, ...this.props.draftValues}}));
      }
        if (prevProps.preferredLocale === this.props.preferredLocale) {
            return;
        }

        // Update the error messages if the language changes
        this.validate(this.state.inputValues);
    }
```
4. We can pass the values for `draftValues` as props in `Form` component.
5. As we don't want to save the input values from draft, we won't pass `shouldSaveDraft` prop in form inputs.

### What alternative solutions did you explore? (Optional)
To avoid any regressions, we can add two props to the `Form` component.
1. `intialValues` - a key-value pair/object of `[inputId]:<inital_value>`
2. `shouldInitializeWithValues` - If this prop is passed we can conditionally initialize the form input values.

If we implement this approach, in future if such issue occurs this can be used.

Looking forward to your feedbacks.


https://github.com/Expensify/App/assets/27822551/39be3e63-a9a2-4c1b-aa6e-46d78fd88e78

**Comment by abdulrahuman5196** [comment]
@akinwale's proposal here https://github.com/Expensify/App/issues/19460#issuecomment-1560405178 looks good and works well.
And as to note its not just description affected, task title is also affected by same issue. We should fix both here since its on the same page and @akinwale's proposal could be applied the same way for title as well.

🎀👀🎀
C+ Reviewed

cc: @tgolen

**Comment by dukenv0307** [comment]
@abdulrahuman5196 Can you check my [proposal](https://github.com/Expensify/App/issues/19460#issuecomment-1562323860) here? My proposal mentioned both description and title bug and @akinwale proposal it's unnecessary to use `defaultValue` if use `value` props and `onChangeText` should replace by `onValueChange` when `TextInput` is wrapped by `Form`

**Comment by abdulrahuman5196** [comment]
@dukenv0307  Thank you for the proposal and reaching out. I did go through all the proposal's at this time before approving.
Both the proposals are almost same. I agree, Your proposal included further concerns of title also needs change and others as mentioned here https://github.com/Expensify/App/issues/19460#issuecomment-1565542059.
But IMO, these are minor improvements which could have easily solved during PR phase. So I have approved the proposal which came first since both are almost same.

**Comment by BhuvaneshPatil** [comment]
@abdulrahuman5196 can you please provide feedback and any improvements to my [proposal](https://github.com/Expensify/App/issues/19460#issuecomment-1564852387).
The reason behind following the approach is we are not using state to store the values anywhere in the code when Form component is used.

**Comment by abdulrahuman5196** [comment]
@BhuvaneshPatil  Thank you reaching out. 
The first part of using task from Onyx is same in all proposals.
The second problem is that Form with TextInput currently won't change the value if the defaultValue is changed. This is a known issue and already have a conversation here - https://github.com/Expensify/App/pull/15672/files#r1139341904 (The conversation is not solved yet). 
But I am not aligned to use draftValues to solve this problem, since it's purpose is to store draft values in form and show it again. IMO using the value prop of TextInput directly like the accepted proposal is more apt in this case.

**Comment by tienifr** [comment]
@abdulrahuman5196 What's about my proposal https://github.com/Expensify/App/issues/19460#issuecomment-1564159075. I don't think adding a new value and onChangeText is a good idea, we already have the logic to render value and handle onChangeText in Form.js, I also wonder if user's typing, the task is updated by pusher(update task from other devices), so the input will be changed. That why I just want to update the input value when setTitleValue or setDescriptionValue is executed.

**Comment by abdulrahuman5196** [comment]
Hi @tienifr , Thank you for the proposal. But I am not aligned with the idea of adding a new field of `shouldSync` just to solve this issue. 

>I don't think adding a new value and onChangeText is a good idea, we already have the logic to render value and handle onInputChange in Form.js

I don't think its forbidden to use value or onChangeText(onValueChange in Form) in Forms, infact Forms expose these params - https://github.com/Expensify/App/blob/main/contributingGuides/FORMS.md. But in most cases you won't need it since Form has a default behaviour and its handled internally by Form. But for our current issue we need more than the default behaviour. 

> I also wonder if user's typing, the task is updated by pusher

I don't think we should worry more about this. These are just draft values stored in onyx.

**Comment by abdulrahuman5196** [comment]
Copy pasting from https://github.com/Expensify/App/issues/19460#issuecomment-1565520844 so that it is not missed in the discussions.


@akinwale's proposal here https://github.com/Expensify/App/issues/19460#issuecomment-1560405178 looks good and works well. Root cause is also correct.
And as to note its not just description affected, task title is also affected by same issue. We should fix both here since its on the same page and @akinwale's proposal could be applied the same way for title as well.

🎀👀🎀
C+ Reviewed

cc: @tgolen

**Comment by situchan** [comment]
I don't think this is worth fixing. Step 7 means that user wants to cancel current edition.
After enter title and close modal and restart creating new task, title and description should be reset.
But after this issue fix, old input values will still remain, even after refreshing app. This is not expected behavior I believe.

**Comment by abdulrahuman5196** [comment]
>After enter title and close modal and restart creating new task, title and description should be reset.
But after this issue fix, old input values will still remain, even after refreshing app. This is not expected behavior I believe.

@situchan  I don't think this will happen. We always clear out the onyx task information before navigating to new task flow even now. Could you kindly let us know on how this issue could happen if you have otherwise information?

**Comment by situchan** [comment]
http://localhost:8080/new/task/details
@abdulrahuman5196 please visit this page directly from address bar

**Comment by abdulrahuman5196** [comment]
Got it @situchan. But I am still aligned to solving this issue. We could also actually solve this concern as part of this issue as well. Anyways I will check with others here well.

**Summarizing for others on the concern.**
Currently after we go to new task page where we assign people, if we refresh the page it will show the already entered information. Same way if we directly go that page via url. But this only can happen if you refresh or directly go via url in web.

https://github.com/Expensify/App/assets/46707890/759af2a3-b767-4f7e-b035-3c88f8835251

This is due to the fact we only clear task information when navigate the new task page. 

So I think the question are,

@tgolen **The same could happen for the page which we are trying to fix this issue. When we refresh or go to url directly, should we show already entered/old information or it should be cleared out?**


Note: I am holding off my approval here https://github.com/Expensify/App/issues/19460#issuecomment-1566187185 to favor the answer to the above question.

**Comment by tgolen** [comment]
I think it is a pleasant UX that it keeps the data that was previously
entered.

**Comment by akinwale** [comment]
> I think it is a pleasant UX that it keeps the data that was previously entered.

Based on this, I suppose we can move forward with my [proposal](https://github.com/Expensify/App/issues/19460#issuecomment-1560405178) as initially [reviewed](https://github.com/Expensify/App/issues/19460#issuecomment-1565520844). Thanks.

cc: @abdulrahuman5196

**Comment by abdulrahuman5196** [comment]
Yes. 

@akinwale's proposal here https://github.com/Expensify/App/issues/19460#issuecomment-1560405178 looks good and works well. Root cause is also correct.
And as to note its not just description affected, task title is also affected by same issue. We should fix both here since its on the same page and @akinwale's proposal could be applied the same way for title as well.
If possible we could also clear the task info when user manually click on the close button, should be a minor change IMO.

🎀👀🎀
C+ Reviewed

cc: @tgolen

**Comment by tgolen** [comment]
Unfortunately, I need to reassign this one due to going on extended leave.

**Comment by greg-schroeder** [comment]
@aldo-expensify thanks for stepping in for Tim! can you take a look at the C+ reviewed proposal [here](https://github.com/Expensify/App/issues/19460#issuecomment-1567953825) and let us know if we're good to proceed?

**Comment by akinwale** [comment]
@aldo-expensify @abdulrahuman5196 I have applied to the Upwork job and the PR is ready for review. Thanks.

**Comment by aldo-expensify** [comment]
Thanks!

**Comment by greg-schroeder** [comment]
C+ review complete, looks like PR is waiting on @aldo-expensify for a final review

**Comment by greg-schroeder** [comment]
Merged, awaiting deploy to prod

**Comment by greg-schroeder** [comment]
Offers sent to @BhuvaneshPatil @akinwale @abdulrahuman5196

**Comment by BhuvaneshPatil** [comment]
Accepted
